### PR TITLE
Algebraic data types

### PR DESCRIPTION
**Pull request by haskellcamargo**
First of all, sorry for the too large pull-request.
This pull-request provides an implementation of algebraic data types based in Haskell. Please note that we still don't have parametric polymorphism. This will come in another branch soon.

We have now:

- `data` declarations
- All primitive types are now written in Quack itself
- Type checker module is isolated
- Support to kinds of types
- Effects for mutability written in Quack itself
- Colored types in the REPL

Well, how important is this? We can model all sort of computations using types and this is the base support for `behavior` and `impl` later. 

Some examples:

```haskell
data Boolean :- True or False

-- Like Haskell does
data List(a) :- Nil or Cons(a, List(a))

-- Implement arithmetic from scratch!
data Natural :- Zero or Next(Natural)

fn &(+)(Zero, b :: Natural) :- value
fn &(+)(Next(a), b)) :- Next(&(+)(a, b))
```

This makes pattern matching possible.
This makes monads possible.
This makes imperative programming via effects possible.

### CODE REVIEW

**Comment by luizperes** [comment]
Travis is failing

**Comment by haskellcamargo** [comment]
@luizperes and it will keep failing because the Hy interpreter was being installed via `pip`, but some change happened and, since then, all builds are failing. The solution for that is removing Hy from the project. For know, we will need to ignore the build. Later, I'll port the test tool to Quack itself. I'm writing a small interpreter to allow testing code in the REPL and, after that, we'll be able to finally write programs in Quack.

File: src/Main.php [hunk 1]
```diff
@@ -39,26 +39,42 @@
...
-            }
-
-            if (!$disable_typechecker) {
-                $parser->ast->runTypeChecker();
+            // Prepend Prelude
+            $prelude = file_get_contents(realpath(dirname(__FILE__) . '/../lib/prelude.qk'));
+            $source = $prelude . PHP_EOL . file_get_contents($file);
+            $script = compile($source, $scope);
+            // Output only provided source instead of all prelude
+            // TODO: This is very ugly. We need to start thinking about modules and export
```
Comment (luizperes) [inline src/Main.php:72]: yep.

File: src/Main.php [hunk 2]
```diff
@@ -70,6 +86,8 @@
 
 $console = new Console(STDIN, STDOUT, STDERR);
 $console->subscribe([
```
Comment (luizperes) [inline src/Main.php:88]: I dont think that this should be here as it is related to the REPL. The main file should be used only to "compile"  (and maybe start the repl), right? Right now it seems that the `Main.php` is the `REPL` itself.

File: src/lexer/Lexer.php [hunk 3]
```diff
@@ -45,8 +45,6 @@ public function __construct($input)
         $this->peek  = $input[0];
 
         // Reserve keywords
-        $this->reserve(new Word(Tag::T_TRUE, 'true'));
```
Comment (luizperes) [inline src/lexer/Lexer.php:48]: good one

File: src/pretty/CliColorizer.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
+<?php
```
Comment (luizperes) [inline src/pretty/CliColorizer.php:1]: why not `cli/pretty`?

File: src/scope/Meta.php [hunk 5]
```diff
@@ -22,9 +22,9 @@
 
 class Meta
 {
-    const M_REF_COUNT = 'refcount';
-    const M_TYPE      = 'type';
-    const M_LABEL     = 'label'; // implicit meta labels
+    const M_REF_COUNT        = 'refcount';
```
Comment (luizperes) [inline src/scope/Meta.php]: not sure why these spaces were added here

**Comment by luizperes** [comment]
Hi @haskellcamargo, Quack has been changing a lot and I see that many many tests are needed in order to keep its quality. Could you update the TODO file so someone can write them? Thanks! :) (even though the tests are currently broken)

**Comment by haskellcamargo** [comment]
@luizperes well, I've ported `run-tests.hy` from LISP to Python, so we can keep the current tests and merge this one. :D

**Comment by luizperes** [comment]
Sounds good! Please just check my considerations and if it all good, you can merge it :) will approve it

**Review by luizperes — APPROVED** [review]

File: src/scope/Meta.php (hunk 5 above)
Comment (haskellcamargo) [reply src/scope/Meta.php]: Because I've added and removed new meta tags several times. I'll strip them.

File: src/pretty/CliColorizer.php (hunk 4 above)
Comment (haskellcamargo) [reply src/pretty/CliColorizer.php:1]: This will be refactored later. I promise.

File: src/Main.php (hunk 2 above)
Comment (haskellcamargo) [reply src/Main.php:88]: `Main` itself is a workaround now, including module loading. This will be fixed together with the module system.

**Pull request merged**